examples/computer_vision/byol_pytorch/model_def.py

```diff
@@ -0,0 +1,367 @@
+        )
+        self.cls_loss_reducers = [
+            cast(
+                AvgReducer,
+                self.context.wrap_reducer(
+                    AvgReducer(), f"cls_loss_{i}", for_training=False
+                ),
+            )
+            for i in range(len(self.hparams.classifier.learning_rates))
+        ]
+
+    def build_callbacks(self) -> Dict[str, PyTorchCallback]:
+        return {"classifier_train": ClassifierTrainCallback(trial=self)}
+
+    def build_training_data_loader(self) -> DataLoader:
+        train_dataset = build_dataset(
+            self.data_config, self.download_dir, split=DatasetSplit.TRAIN
+        )
+        # In order to ensure distributed training shards correctly, we round the dataset size
+        # down to be a multiple of the global batch size.
+        # See comment here for more details:
+        #   https://github.com/determined-ai/determined/blob/b3a34baa7dcca788a090120a17f9a4f8dc1a4184/harness/determined/pytorch/samplers.py#L77
+        rounded_length = (
+            len(train_dataset) // self.context.get_global_batch_size()
+        ) * self.context.get_global_batch_size()
+        train_dataset = torch.utils.data.Subset(train_dataset, range(rounded_length))
+        return DataLoader(
+            train_dataset,
+            batch_size=self.context.get_per_slot_batch_size(),
+            pin_memory=True,
+            num_workers=self.data_config.num_workers,
+            shuffle=True,
+            drop_last=True,
+        )
+
+    def _build_cls_training_data_loader(self) -> DataLoader:
+        cls_train_dataset = build_dataset(
+            self.data_config, self.download_dir, DatasetSplit.CLS_TRAIN
+        )
+        rounded_length = (
+            len(cls_train_dataset) // self.context.get_global_batch_size()
+        ) * self.context.get_global_batch_size()
+        cls_train_dataset = torch.utils.data.Subset(
+            cls_train_dataset, range(rounded_length)
+        )
+        return DataLoader(
+            cls_train_dataset,
+            batch_size=self.context.get_per_slot_batch_size(),
+            pin_memory=True,
+            num_workers=self.data_config.num_workers,
+            shuffle=True,
+            drop_last=True,
+        )
+
+    def build_validation_data_loader(self) -> DataLoader:
+        # BYOL performs validation in two steps:
+        # - A subset of the training data is used to select an optimal LR for the classifier.
+        # - Final results are reported on the test / validation data.
+        # We combine these two datasets, and then use a custom reducer to calculate the final result.
+        lr_val_dataset = build_dataset(
+            self.data_config,
+            self.download_dir,
+            DatasetSplit.CLS_VALIDATION,
+        )
+        test_dataset = build_dataset(
+            self.data_config, self.download_dir, DatasetSplit.TEST
+        )
+        combined = JointDataset([lr_val_dataset, test_dataset], ["lr_val", "test"])
```

Comment:
are the lr_val_dataset and test_dataset always the same size?

Reply:
No, in general they are different sizes.